### PR TITLE
fix scope and resource middleware data access

### LINKED ISSUE #1790: Cannot access app_data in middleware of Scope
## Simple Use Case
```rust
...
web::scope("/api")
    .app_data(web::Data::new(1usize))
    .wrap_fn(|req, srv| {
        let i = req.app_data::<usize>();
        println!("i {} ", i.is_some()); // outputs "i false"
        let fut = srv.call(req);

        async move {
            let res = fut.await?;
            Ok(res)
        }
    })
...
```
## Expected Behavior
Ability to access the request's `app_data` within the middleware (`wrap` / `wrap_fn`) of a `Scope`.

## Current Behavior
Looking at the example code above, you would expect that `app_data` from the request can be retrieved, but it does not work.
If you move the setting of `app_data` outside the scope (ex. on `App::new().app_data(data)`) the data is accessible from within the middleware.

## Possible Solution
Spoke with @robjtede on the Gitter channel and he advised me to submit this issue. I am currently looking into debugging it myself, but I am new to the code.

One thought that I have is that in the use case above you can't get the data because `ScopeService` doesn't invoke `call()` until after trying to get the data, since on `ScopeService::call()` it stuffs its data into the request.

* Rust Version (I.e, output of `rustc -V`): rustc 1.50.0-nightly (825637983 2020-11-18)
* Actix Web Version: actix-web = "3"

### PR DESCRIPTION
**Pull request by robjtede**
## PR Type

Fix

## PR Checklist

- [x] Tests for the changes have been added / updated.
- [x] Documentation comments have been added / updated.
- [x] A changelog entry has been made for the appropriate packages.
- [x] Format code with the latest stable rustfmt.
- [x] (Team) Label with affected crates and semver status.

## Overview
Fixes middleware's access to self-layer data on `Scope`s and `Resource`s using the approach outlined in #2261. Tests ensure that both routes and default services are wrapped with data injection in both cases.

Co-authored by: @aliemjay. Thanks for their work exploring both approaches to this; this one was chosen over #2267 because it more closely matches how data injection works for `App` and does not change public API.

Also technically removes an `IntoServiceFactory` impl for `Resource` which looks like a mistake because it allowed `Resource` to be used as a `default_service` which doesn't make sense. Intentionally not included in changelog.

supersedes #2267
supersedes #2261
closes #1790

### CODE REVIEW

File: src/resource.rs [hunk 1]
```diff
@@ -400,34 +400,28 @@ where
...
             routes: self.routes,
-            app_data: self.app_data.map(Rc::new),
             default: self.default,
         });
 
-        self.endpoint
+        let resource_data = self.app_data.map(Rc::new);
+
+        // wraps endpoint service (including middleware) call and injects app data for this scope
+        let endpoint = apply_fn_factory(self.endpoint, move |mut req: ServiceRequest, srv| {
```
Comment (ibraheemdev) [inline src/resource.rs:411]: Is the type annotation (on `req`) needed?

Comment (robjtede) [reply src/resource.rs:411]: yep

```
454 |         let endpoint = apply_fn_factory(self.endpoint, move |mut req, srv| {
    |                                                              ^^^^^^^ consider giving this closure parameter a type
```

**Review by aliemjay — APPROVED** [review]
LGTM.

Have you considered not to add a middleware in the common case where `app_data == None`? I have this in #2264. Maybe micro-optimization?

**Comment by robjtede** [comment]
> Have you considered not to add a middleware in the common case where app_data == None? I have this in #2264. Maybe micro-optimization?

Not concerned about that now, one branch vs one nested stack call is not worth the added complexity. If we want it can be added later without problems.

**Comment by aliemjay** [comment]
> Also technically removes an `IntoServiceFactory` impl for `Resource` which looks like a mistake because it allowed `Resource` to be used as a `default_service` which doesn't make sense. Intentionally not included in changelog.

I already had this in mind. But my suggested solution is more radical: `Route` shouldn't have `impl ServiceFactory` because it makes no use of the defined guards. `Route::to` and `::default_service` should have the same signature with a new super trait with a blanket impl for `IntoServiceFactory` and `Handler`. This would also revert #2262.

**Pull request merged**